File: backend/routes/election.js
const express = require('express');
const router = express.Router();
const pool = require('../config/db');
const { verifyAdmin } = require('../middleware/auth');

function computeStatus(start, end, now) {
  if (!start || !end) return 'not_configured';
  if (now < new Date(start)) return 'not_started';
  if (now <= new Date(end)) return 'open';
  return 'closed';
}

// Public: current election title, window, and status
router.get('/', async (req, res) => {
  try {
    const [[row]] = await pool.query('SELECT title, start_time, end_time FROM election_settings WHERE id = 1');
    const now = new Date();
    const status = computeStatus(row.start_time, row.end_time, now);
    res.json({
      title: row.title,
      start_time: row.start_time,
      end_time: row.end_time,
      server_time: now.toISOString(),
      status
    });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: 'Could not load election settings' });
  }
});

// Admin: update title and/or voting window
router.put('/', verifyAdmin, async (req, res) => {
  const { title, start_time, end_time } = req.body;
  if (!title || !start_time || !end_time) {
    return res.status(400).json({ message: 'Title, start time, and end time are all required' });
  }
  if (new Date(end_time) <= new Date(start_time)) {
    return res.status(400).json({ message: 'End time must be after start time' });
  }

  try {
    await pool.query(
      'UPDATE election_settings SET title = ?, start_time = ?, end_time = ? WHERE id = 1',
      [title, new Date(start_time), new Date(end_time)]
    );
    res.json({ message: 'Election schedule updated' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: 'Could not update election settings' });
  }
});

module.exports = router;

File: backend/app/routers/election.py
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException
from ..database import get_conn
from ..schemas import ElectionUpdate
from ..deps import get_current_admin
from ..auth_utils import iso

router = APIRouter(prefix="/api/election", tags=["election"])


def compute_status(start, end, now):
    if not start or not end:
        return "not_configured"
    if now < start:
        return "not_started"
    if now <= end:
        return "open"
    return "closed"


def parse_dt(value: str) -> datetime:
    # <input type="datetime-local"> omits seconds ("YYYY-MM-DDTHH:MM");
    # pad them in so fromisoformat works on every supported Python version.
    if len(value) == 16:
        value += ":00"
    return datetime.fromisoformat(value)


@router.get("")
def get_election():
    conn = get_conn()
    try:
        with conn.cursor() as cur:
            cur.execute("SELECT title, start_time, end_time FROM election_settings WHERE id = 1")
            row = cur.fetchone()
        now = datetime.now()
        status = compute_status(row["start_time"], row["end_time"], now)
        return {
            "title": row["title"],
            "start_time": iso(row["start_time"]),
            "end_time": iso(row["end_time"]),
            "server_time": iso(now),
            "status": status,
        }
    finally:
        conn.close()


@router.put("")
def update_election(payload: ElectionUpdate, admin=Depends(get_current_admin)):
    start = parse_dt(payload.start_time)
    end = parse_dt(payload.end_time)
    if end <= start:
        raise HTTPException(status_code=400, detail="End time must be after start time")

    conn = get_conn()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "UPDATE election_settings SET title=%s, start_time=%s, end_time=%s WHERE id=1",
                (payload.title, start, end),
            )
        conn.commit()
        return {"message": "Election schedule updated"}
    finally:
        conn.close()
